google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystemProvider.java

```diff
@@ -515,40 +527,50 @@ public void copy(Path source, Path target, CopyOption... options) throws IOExcep
       throw new CloudStoragePseudoDirectoryException(toPath);
     }
 
-    try {
-      if (wantCopyAttributes) {
-        BlobInfo blobInfo = storage.get(fromPath.getBlobId());
-        if (null == blobInfo) {
-          throw new NoSuchFileException(fromPath.toString());
-        }
-        if (!setCacheControl) {
-          tgtInfoBuilder.setCacheControl(blobInfo.getCacheControl());
-        }
-        if (!setContentType) {
-          tgtInfoBuilder.setContentType(blobInfo.getContentType());
+    final CloudStorageRetryHandler retryHandler = new CloudStorageRetryHandler(CloudStorageUtil.getMaxChannelReopensFromPath(source));
+    // Loop will terminate via an exception if all retries are exhausted
+    while (true) {
+      try {
+        if ( wantCopyAttributes ) {
+          BlobInfo blobInfo = storage.get(fromPath.getBlobId());
+          if ( null == blobInfo ) {
+            throw new NoSuchFileException(fromPath.toString());
+          }
+          if ( !overrideCacheControl ) {
+            tgtInfoBuilder.setCacheControl(blobInfo.getCacheControl());
+          }
+          if ( !overrideContentType ) {
+            tgtInfoBuilder.setContentType(blobInfo.getContentType());
+          }
+          if ( !overrideContentEncoding ) {
+            tgtInfoBuilder.setContentEncoding(blobInfo.getContentEncoding());
+          }
+          if ( !overrideContentDisposition ) {
+            tgtInfoBuilder.setContentDisposition(blobInfo.getContentDisposition());
+          }
+          tgtInfoBuilder.setAcl(blobInfo.getAcl());
+          tgtInfoBuilder.setMetadata(blobInfo.getMetadata());
         }
-        if (!setContentEncoding) {
-          tgtInfoBuilder.setContentEncoding(blobInfo.getContentEncoding());
+
+        BlobInfo tgtInfo = tgtInfoBuilder.build();
+        Storage.CopyRequest.Builder copyReqBuilder =
+                Storage.CopyRequest.newBuilder().setSource(fromPath.getBlobId());
+        if (wantReplaceExisting) {
+          copyReqBuilder = copyReqBuilder.setTarget(tgtInfo);
+        } else {
+          copyReqBuilder = copyReqBuilder.setTarget(tgtInfo, Storage.BlobTargetOption.doesNotExist());
         }
-        if (!setContentDisposition) {
-          tgtInfoBuilder.setContentDisposition(blobInfo.getContentDisposition());
+        CopyWriter copyWriter = storage.copy(copyReqBuilder.build());
+        copyWriter.getResult();
```

Comment:
BUGBUG: There needs to be a `break;` after this line to end the enclosing loop when the copy is successful.